pkg/querier/queryrange/limits.go

```diff
@@ -528,40 +507,30 @@ func (rt limitedRoundTripper) RoundTrip(r *http.Request) (*http.Response, error)
 		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, ErrMaxQueryParalellism.Error())
 	}
 
-	for i := 0; i < parallelism; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			for {
-				select {
-				case w := <-intermediate:
-					resp, err := rt.do(w.ctx, w.req)
-					w.result <- result{response: resp, err: err}
-				case <-ctx.Done():
-					return
-				}
-			}
-		}()
-	}
+	sem := semaphore.NewWeighted(int64(parallelism))
 
 	response, err := rt.middleware.Wrap(
 		queryrangebase.HandlerFunc(func(ctx context.Context, r queryrangebase.Request) (queryrangebase.Response, error) {
-			w := newWork(ctx, r)
-			select {
-			case intermediate <- w:
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			}
-			select {
-			case response := <-w.result:
-				return response.response, response.err
-			case <-ctx.Done():
-				return nil, ctx.Err()
+			// This inner handler is called multiple times by
+			// sharding outer middlewares such as the downstreamer.
+			// The number of concurrent calls to `next.RoundTrip` should
+			// be limited, because the downstream calls can be in
+			// the thousands.
+			// Note: It is the responsibility of the caller to run
+			// the handler in parallel.
+			if err := sem.Acquire(ctx, int64(1)); err != nil {
+				return nil, fmt.Errorf("could not acquire work: %w", err)
 			}
+			defer sem.Release(int64(1))
+
+			return rt.do(ctx, r)
 		})).Do(ctx, request)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO(karsten): Do we have to wait for all work being done
+	// sem.Acquire(ctx, int64(parallelism))
```

Comment:
I don't think so since we are waiting in line 512 but we could just to be safe.

Reply:
Agree, it does not seem to be necessary.